In [18]:
#import dependencies 
import pandas as pd 

In [19]:
#file path and read file as Pandas Dataframe 
raw_data = "../HeroesOfPymoli/purchase_data.csv"
purchase_data_df = pd.read_csv(raw_data)
purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

### Player Count

* Total Number of Players

In [20]:
player_count = len(purchase_data_df["SN"].unique())
player_count

576

### Purchasing Analysis (Total)

* Number of Unique Items
* Average Purchase Price
* Total Number of Purchases
* Total Revenue

In [21]:
#Number of unique items.
unique_item = len(purchase_data_df["Item Name"].unique())
unique_item

179

In [22]:
#Average Purchase Price 
average_purchase = purchase_data_df["Price"].mean()
average_purchase

3.050987179487176

In [23]:
#Total number of purchases
total_purchases = len(purchase_data_df)
total_purchases

780

In [24]:
# Total Revenue.
total_revenue= purchase_data_df["Price"].sum()
total_revenue.round(2)

2379.77

In [25]:
#Make a data frame out of a dictionary of the new values 
purchasing_analysis = pd.DataFrame({"Number of Unique Items":[unique_item],
                                             "Average Purchase Price":[average_purchase],
                                            "Total Number of Purchases":[total_purchases],
                                             "Total Revenue":[total_revenue]})
purchasing_analysis

Number of Unique Items  Average Purchase Price  Total Number of Purchases  \
0                     179                3.050987                        780   

   Total Revenue  
0        2379.77

### Gender Demographics

* Percentage and Count of Male Players
* Percentage and Count of Female Players
* Percentage and Count of Other / Non-Disclosed

In [26]:
#Count of Male Players, Female Players and Other/Non-Disclosed 
gender_count = purchase_data_df.groupby("Gender")["SN"].nunique()
gender_count.head()


Gender
Female                    81
Male                     484
Other / Non-Disclosed     11
Name: SN, dtype: int64

In [27]:
#Percentage of Male Players,Female Players and Other/Non-Disclosed
gender_percent = (gender_count/player_count)*100
gender_percent.round(2)

Gender
Female                   14.06
Male                     84.03
Other / Non-Disclosed     1.91
Name: SN, dtype: float64

In [28]:
##Make a data frame out of a dictionary of the new values

gender_demographics = pd.DataFrame({"Gender Count": gender_count,
                                    "Gender Percentage":gender_percent})
gender_demographics

Gender Count  Gender Percentage
Gender                                                
Female                           81          14.062500
Male                            484          84.027778
Other / Non-Disclosed            11           1.909722

### Purchasing Analysis (Gender)

* The below each broken by gender
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
  * Average Purchase Total per Person by Gender

In [47]:
#Purchase Count by gender
gender_purchases = purchase_data_df.groupby("Gender")["Item Name"]
gender_p_count = gender_purchases.count()
gender_p_count

Gender
Female                   113
Male                     652
Other / Non-Disclosed     15
Name: Item Name, dtype: int64

In [30]:
#Average Purchase Price by gender 
gender_average_p = purchase_data_df.groupby("Gender")["Price"].mean()
gender_average_p.round(2)

Gender
Female                   3.20
Male                     3.02
Other / Non-Disclosed    3.35
Name: Price, dtype: float64

In [31]:
#Total Purchase Value by gender 
gender_total_p = purchase_data_df.groupby("Gender")["Price"].sum()
gender_total_p

Gender
Female                    361.94
Male                     1967.64
Other / Non-Disclosed      50.19
Name: Price, dtype: float64

In [32]:
#Average Purchase Total Per Person by gender
aptpp= gender_total_p/gender_count
aptpp.round(2)

Gender
Female                   4.47
Male                     4.07
Other / Non-Disclosed    4.56
dtype: float64

In [48]:
# Purchasing analysis DataFrame by gender.

gender_analysis = pd.DataFrame({"Purchase Count":gender_p_count, 
                                   "Average Purchase Price":gender_average_p,
                                   "Total Purchase Value":gender_total_p,
                                   "Average Purchase Total Per Person":aptpp})
gender_analysis

Purchase Count  Average Purchase Price  \
Gender                                                          
Female                            113                3.203009   
Male                              652                3.017853   
Other / Non-Disclosed              15                3.346000   

                       Total Purchase Value  Average Purchase Total Per Person  
Gender                                                                          
Female                               361.94                           4.468395  
Male                                1967.64                           4.065372  
Other / Non-Disclosed                 50.19                           4.562727

### Age Demographics

* The below each broken into bins of 4 years (i.e. &lt;10, 10-14, 15-19, etc.)
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
  * Average Purchase Total per Person by Age Group

In [49]:
#Create bins: <10, 10-14, 15-19, 20-24, 25-29, 30-34, 35-39 >39.

bins = [0,10,15,20,25,30,35,40, 45]
age_ranges = ["<10", "10-14","15-19", "20-24", "25-29", "30-34", "35-39", ">=40"]

In [50]:
# Cut purchase data and place the ages into bins
purchase_data_df["Age Range"] = pd.cut(purchase_data_df["Age"], bins, labels=age_ranges, include_lowest=True)
purchase_data_df

Purchase ID             SN  Age  Gender  Item ID  \
0              0        Lisim78   20    Male      108   
1              1    Lisovynya38   40    Male      143   
2              2     Ithergue48   24    Male       92   
3              3  Chamassasya86   24    Male      100   
4              4      Iskosia90   23    Male      131   
..           ...            ...  ...     ...      ...   
775          775     Aethedru70   21  Female       60   
776          776         Iral74   21    Male      164   
777          777     Yathecal72   20    Male       67   
778          778        Sisur91    7    Male      101   
779          779      Ennrian78   24    Male       50   

                                     Item Name  Price Age Range  
0    Extraction, Quickblade Of Trembling Hands   3.53     15-19  
1                            Frenzied Scimitar   1.56     35-39  
2                                 Final Critic   4.88     20-24  
3                                  Blindscythe   3.27     20-24  
4                                         Fury   1.44     20-24  
..                                         ...    ...       ...  
775                                       Wolf   3.54     20-24  
776                           Exiled Doomblade   1.63     20-24  
777      Celeste, Incarnation of the Corrupted   3.46     15-19  
778                               Final Critic   4.19       <10  
779                                       Dawn   4.60     20-24  

[780 rows x 8 columns]

In [51]:
# Purchase count by age range.

age_purchase_count = purchase_data_df.groupby("Age Range")["Item Name"]
age_p_count = age_purchase_count.count()
age_p_count

Age Range
<10       32
10-14     54
15-19    200
20-24    325
25-29     77
30-34     52
35-39     33
>=40       7
Name: Item Name, dtype: int64

In [37]:
# Average purchase price by age range.

age_purchase_price= purchase_data_df.groupby("Age Range")["Price"].mean()
age_purchase_price.round(2)

Age Range
<10      3.40
10-14    2.90
15-19    3.11
20-24    3.02
25-29    2.88
30-34    2.99
35-39    3.40
>=40     3.08
Name: Price, dtype: float64

In [38]:
#Total purchase value by age range.

age_total_purchase = purchase_data_df.groupby("Age Range")["Price"].sum()
age_total_purchase

Age Range
<10      108.96
10-14    156.60
15-19    621.56
20-24    981.64
25-29    221.42
30-34    155.71
35-39    112.35
>=40      21.53
Name: Price, dtype: float64

In [39]:
#Average Purchase Total Per Person by Age Group 

age_aptpp = age_total_purchase/player_count
age_aptpp.round(2)

Age Range
<10      0.19
10-14    0.27
15-19    1.08
20-24    1.70
25-29    0.38
30-34    0.27
35-39    0.20
>=40     0.04
Name: Price, dtype: float64

In [52]:
# Age Demographics DataFrame

age_demographics = pd.DataFrame({"Purchase Count":age_p_count,
                            "Average Purchase Price":age_purchase_price,
                            "Total Purchase Value": age_total_purchase,
                            "Average Purchase Total Per Person": age_aptpp})
age_demographics

Purchase Count  Average Purchase Price  Total Purchase Value  \
Age Range                                                                 
<10                    32                3.405000                108.96   
10-14                  54                2.900000                156.60   
15-19                 200                3.107800                621.56   
20-24                 325                3.020431                981.64   
25-29                  77                2.875584                221.42   
30-34                  52                2.994423                155.71   
35-39                  33                3.404545                112.35   
>=40                    7                3.075714                 21.53   

           Average Purchase Total Per Person  
Age Range                                     
<10                                 0.189167  
10-14                               0.271875  
15-19                               1.079097  
20-24                               1.704236  
25-29                               0.384410  
30-34                               0.270330  
35-39                               0.195052  
>=40                                0.037378

### Top Spenders

* Identify the the top 5 spenders in the game by total purchase value, then list (in a table):
  * SN
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value

In [41]:

players_purchase_count = purchase_data_df.groupby("SN").count()["Price"].rename("Purchase Count")
players_average_price = purchase_data_df.groupby("SN").mean()["Price"].rename("Average Purchase Price")
players_total = purchase_data_df.groupby("SN").sum()["Price"].rename("Total Purchase Value")

#Convert to DataFrame.

total_users = pd.DataFrame({"Purchase Count":players_purchase_count,
                                   "Average Purchase Price": players_average_price,
                                   "Total Purchase Value": players_total})
total_users.head()

Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                        
Adairialis76               1                    2.28                  2.28
Adastirin33                1                    4.48                  4.48
Aeda94                     1                    4.91                  4.91
Aela59                     1                    4.32                  4.32
Aelaria33                  1                    1.79                  1.79

In [42]:
# Sort table to show the top five spenders.
top_five = total_users.sort_values("Total Purchase Value", ascending=False)
top_five.head(5)

Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                       
Lisosia93                 5                3.792000                 18.96
Idastidru52               4                3.862500                 15.45
Chamjask73                3                4.610000                 13.83
Iral74                    4                3.405000                 13.62
Iskadarya95               3                4.366667                 13.10

### Most Popular Items

* Identify the 5 most popular items by purchase count, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value

In [43]:
# Total items purchases analysis.

items_purchase_count = purchase_data_df.groupby(["Item ID", "Item Name"]).count()["Price"].rename("Purchase Count")
items_average_price = purchase_data_df.groupby(["Item ID", "Item Name"]).mean()["Price"].rename("Average Purchase Price")
items_value_total = purchase_data_df.groupby(["Item ID", "Item Name"]).sum()["Price"].rename("Total Purchase Value")

# Convert to DataFrame

items_purchased = pd.DataFrame({"Purchase Count":items_purchase_count,
                                   "Item Price":items_average_price,
                                   "Total Purchase Value":items_value_total})

items_purchased.head()

Purchase Count  Item Price  Total Purchase Value
Item ID Item Name                                                           
0       Splinter                         4        1.28                  5.12
1       Crucifer                         3        3.26                  9.78
2       Verdict                          6        2.48                 14.88
3       Phantomlight                     6        2.49                 14.94
4       Bloodlord's Fetish               5        1.70                  8.50

In [44]:
# Sort table to show the five the most popular items.

popular_items = items_purchased.sort_values("Purchase Count", ascending=False)
popular_items.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                      Item Price  \
Item ID Item Name                                                  
178     Oathbreaker, Last Hope of the Breaking Storm        4.23   
145     Fiery Glass Crusader                                4.58   
108     Extraction, Quickblade Of Trembling Hands           3.53   
82      Nirvana                                             4.90   
19      Pursuit, Cudgel of Necromancy                       1.02   

                                                      Total Purchase Value  
Item ID Item Name                                                           
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76  
145     Fiery Glass Crusader                                         41.22  
108     Extraction, Quickblade Of Trembling Hands                    31.77  
82      Nirvana                                                      44.10  
19      Pursuit, Cudgel of Necromancy                                 8.16

### Most Profitable Items

* Identify the 5 most profitable items by total purchase value, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value

In [45]:
# Sort table to show the five the most profitable items.

profitable_items = items_purchased.sort_values("Total Purchase Value", ascending=False)
profitable_items.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                      Item Price  \
Item ID Item Name                                                  
178     Oathbreaker, Last Hope of the Breaking Storm        4.23   
82      Nirvana                                             4.90   
145     Fiery Glass Crusader                                4.58   
92      Final Critic                                        4.88   
103     Singed Scalpel                                      4.35   

                                                      Total Purchase Value  
Item ID Item Name                                                           
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76  
82      Nirvana                                                      44.10  
145     Fiery Glass Crusader                                         41.22  
92      Final Critic                                                 39.04  
103     Singed Scalpel                                               34.80

Analysis:



1. The age range that made the greatest amount of purchases was the age range of 20-24. This age range also had the highest total purchase value as well as highest average purchase total per person.
2. Item #178, the "Oathbreaker,Last Hope of the Breaking Storm" is the number one most popular item as well as the number one most profitable item. 
3. Males overwhelmingly made more purchases than females with a ratio of 652:113, however this could also be because there are more males in this data than females with a ratio of 484:81. Thus, when comparing the purchasing habits between females and males, the average purchase price and average purchse total per person may be more accurate to analyze. These parameters tell us that the average purchase price between females and males were very similar with values of $3.20 and $3.02,respectively. 